path_decrypt.go

```diff
@@ -119,11 +123,19 @@ func (b *backend) pathDecryptWrite(ctx context.Context, req *logical.Request, d
 
 	// Lookup the key so we can determine the type of decryption (symmetric or
 	// asymmetric).
-	ck, err := kmsClient.GetCryptoKey(ctx, &kmspb.GetCryptoKeyRequest{
-		Name: k.CryptoKeyID,
-	})
-	if err != nil {
-		return nil, errwrap.Wrapf("failed to get underlying crypto key: {{err}}", err)
+	var ck *kmspb.CryptoKey
+	cachedCk, found := keysCache.Get(k.CryptoKeyID)
+	if !found {
+		fetchedCk, err := kmsClient.GetCryptoKey(ctx, &kmspb.GetCryptoKeyRequest{
+			Name: k.CryptoKeyID,
+		})
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to get underlying crypto key: {{err}}", err)
+		}
+		_ = keysCache.Add(k.CryptoKeyID, fetchedCk, cache.DefaultExpiration)
+		ck = fetchedCk
+	} else {
+		ck = cachedCk.(*kmspb.CryptoKey)
```

Comment:
I agree the casting is relatively safe since we know the only values that get cached at the moment are valid crypto keys, but just in case we should probably verify the type assertion just in case to fail gracefully if other parts of the code ever cache something invalid:

```suggestion
		castedCk, ok := cachedCk.(*kmspb.CryptoKey)
		if !ok {
			return nil, fmt.Errorf("failed to get underlying crypto key: %q", err)
		}
		ck = castedCk
```

Reply:
> @maxcoulombe changes have been made

Awesome, the changes look good to me. Thanks again for that contribution. 

I'll close the related issue and check with the team to kick-off a release. This should be automatically included in the future Vault 1.16 release as well.